### PR TITLE
Support filtering on bool/scalar fields without evaluator

### PR DESCRIPTION
**Pull request by vvivekiyer**
Fixes the bugs in #8444 and #8487.
Added unit tests.

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java [hunk 1]
```diff
@@ -48,4 +48,13 @@ public boolean isRange() {
...
   }
+
+  /**
+   * Returns true if the filter is a predicate. Returns false otherwise. This logic should mimic FilterContext.Type.
+   *
+   * @return where the filter is a predicate.
+   */
+  public boolean isPredicate() {
+    return this != AND && this != OR && this != NOT;
+  }
```
Comment (richardstartin) [inline pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:59]: This, along with the method `isRange` above it, is a little messy - it would be better to add fields to the enum `_isRange` and `_isPredicate` and set them in the enum initialiser.

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java [hunk 2]
```diff
@@ -2246,6 +2246,19 @@ public void testFlattenAndOr() {
...
+      String query = "SELECT * FROM foo WHERE col1 > 0 AND (col2 AND col3 > 0) AND startsWith(col4, 'myStr')";
+      PinotQuery pinotQuery = CalciteSqlParser.compileToPinotQuery(query);
+      Function functionCall = pinotQuery.getFilterExpression().getFunctionCall();
+      Assert.assertEquals(functionCall.getOperator(), FilterKind.AND.name());
+      List<Expression> operands = functionCall.getOperands();
+      Assert.assertEquals(operands.size(), 4);
+      Assert.assertEquals(operands.get(0).getFunctionCall().getOperator(), FilterKind.GREATER_THAN.name());
+      Assert.assertEquals(operands.get(1).getFunctionCall().getOperator(), FilterKind.EQUALS.name());
+      Assert.assertEquals(operands.get(2).getFunctionCall().getOperator(), FilterKind.GREATER_THAN.name());
+      Assert.assertEquals(operands.get(3).getFunctionCall().getOperator(), FilterKind.EQUALS.name());
```
Comment (siddharthteotia) [inline pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:2259]: (nit) for the part that is being rewritten `startsWith(col4, 'myStr')`, we may want to assert for both LHS and RHS (esp RHS)

So in above test, for the 4th operand we are only checking for operator EQUALS. Would be good to check LHS is startsWith and RHS is true

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java [hunk 3]
```diff
@@ -2254,9 +2267,9 @@ public void testFlattenAndOr() {
       List<Expression> operands = functionCall.getOperands();
       Assert.assertEquals(operands.size(), 4);
       Assert.assertEquals(operands.get(0).getFunctionCall().getOperator(), FilterKind.GREATER_THAN.name());
-      Assert.assertEquals(operands.get(1).getIdentifier().getName(), "col2");
+      Assert.assertEquals(operands.get(1).getFunctionCall().getOperator(), FilterKind.EQUALS.name());
       Assert.assertEquals(operands.get(2).getFunctionCall().getOperator(), FilterKind.GREATER_THAN.name());
-      Assert.assertEquals(operands.get(3).getIdentifier().getName(), "col4");
+      Assert.assertEquals(operands.get(3).getFunctionCall().getOperator(), FilterKind.EQUALS.name());
```
Comment (siddharthteotia) [inline pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:2272]: Same here

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java [hunk 4]
```diff
@@ -2246,6 +2246,19 @@ public void testFlattenAndOr() {
       }
     }
```
Comment (siddharthteotia) [inline pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:2248]: We may also want to add a test where the boolean expression is as expected (e.g `WHERE col1 = true`) and therefore rewrite is a NO-OP  or a mix of expressions where some are rewritten and some are NO-OP?

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java [hunk 5]
```diff
@@ -42,6 +44,73 @@ public PinotQuery rewrite(PinotQuery pinotQuery) {
...
+      }
+    } else if (type == ExpressionType.IDENTIFIER) {
+
+      // Rewrite identifiers  if one of the two conditions are satified:
+      // 1. parent function is empty. Example: "select * from table where col1";
+      // 2. Separator is not a predicate (non predicates are AND, OR, NOT).
+      //    Example: "select * from table where col1 AND col2 > 10";
+      if (parentFunction == null || (EnumUtils.isValidEnum(FilterKind.class, parentFunction.getOperator())
+          && !FilterKind.valueOf(parentFunction.getOperator()).isPredicate())) {
+        Expression currExpression = expression.deepCopy();
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java:102]: In-place modification / rewrite  not possible ?

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java [hunk 6]
```diff
@@ -42,6 +44,73 @@ public PinotQuery rewrite(PinotQuery pinotQuery) {
...
+
+      List<Expression> operands = expression.getFunctionCall().getOperands();
+
+      if (operands.size() > 0) {
+        // unset operands only within the "if" condition because 'no argument' functions expect operand list to
+        // be initialized.
+        expression.getFunctionCall().unsetOperands();
+
+        for (Expression exp : operands) {
+          expression.getFunctionCall().addToOperands(updateBooleanPredicates(exp, expression.getFunctionCall()));
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java:91]: (nit) for shorter / cleaner code: do expression.getFunctionCall() once ?
`Expression parent = expression.getFunctionCall()`

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java [hunk 7]
```diff
@@ -42,6 +44,73 @@ public PinotQuery rewrite(PinotQuery pinotQuery) {
...
+        }
+      }
+    } else if (type == ExpressionType.IDENTIFIER) {
+
+      // Rewrite identifiers  if one of the two conditions are satified:
+      // 1. parent function is empty. Example: "select * from table where col1";
+      // 2. Separator is not a predicate (non predicates are AND, OR, NOT).
+      //    Example: "select * from table where col1 AND col2 > 10";
+      if (parentFunction == null || (EnumUtils.isValidEnum(FilterKind.class, parentFunction.getOperator())
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java:100]: If the particular node is an identifier under the WHERE tree then does it matter if it has a parent or not ?

In both the above examples, as you traverse down the filter tree, you will reach `WHERE col1`, and enter this else branch. We just need to rewrite it . Why should we check that parent should be null or one of AND,. OR, NOT ?

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java (hunk 1 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java]: Done.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java (hunk 6 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java]: I've refactored code and reused variables. Please take a look.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java (hunk 5 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java]: I've refactored code and modified in-place. Please take a look.

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java (hunk 4 above)
Comment (vvivekiyer) [reply pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:2348]: Added the following tests. WHERE clause containing:
1. Only boolean field.
2. Only boolean field (and) boolean field with equals.
3. Only scalar field.
4. Only scalar field (and) scalar field with equals.
5. Only scalar field (and) scalar field with equals and only boolean field (and) boolean field with equals

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java (hunk 7 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java]: In both the above example, when we reach `WHERE col1` we have to rewrite the query.
But consider the example query `WHERE concat(col1, "str")`. In this case we have to not rewrite the query. 

So the logic is that, literals will be re-written only if at least one of the conditions is satisfied:
1. Parent function is null. 
2. Parent function is one of AND, OR, NOT.

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java (hunk 2 above)
Comment (vvivekiyer) [reply pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:2362]: Done.

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java (hunk 3 above)
Comment (vvivekiyer) [reply pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:2381]: Done.

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java [hunk 8]
```diff
@@ -39,13 +41,26 @@ public enum FilterKind {
   TEXT_MATCH,
   JSON_MATCH;
 
+  private static final EnumSet<FilterKind> RANGE_FILTERS = EnumSet.of(GREATER_THAN, GREATER_THAN_OR_EQUAL, LESS_THAN,
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:44]: It is easy to forget updating the EnumSet when one updates the FilterKind. One option may be to move LIKE after BETWEEN and then do:

  public boolean isRange() {
    return this >= GREATER_THAN && this <= RANGE;
  }

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java [hunk 9]
```diff
@@ -39,13 +41,26 @@ public enum FilterKind {
...
+      LESS_THAN_OR_EQUAL, BETWEEN, RANGE);
+  private static final EnumSet<FilterKind> PREDICATE_FILTERS = EnumSet.of(EQUALS, NOT_EQUALS, GREATER_THAN,
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:46]: Same here, it's easy to forget updating PREDICATE_FILTERS when one updates FilterKind. Maybe the ordinal positions of enum in FilterKind can be fixed and ordinal positions used? That way we only have to maintain a single data structure.

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java [hunk 10]
```diff
@@ -39,13 +41,26 @@ public enum FilterKind {
...
-    return this == GREATER_THAN || this == GREATER_THAN_OR_EQUAL || this == LESS_THAN || this == LESS_THAN_OR_EQUAL
-        || this == BETWEEN || this == RANGE;
+    return RANGE_FILTERS.contains(this);
+  }
+
+  /**
+   * Returns true if the filter is a predicate. Returns false otherwise. This logic should mimic FilterContext.Type.
+   *
+   * @return where the filter is a predicate.
+   */
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:62]: A single line comment should be sufficient here :-) Same with comment on isRange function.
/** @return true if the filter is a predicate; otherwise, false. Logic should mimic FilterContext.Type. */

Can the logic be moved to a common place to avoid keeping the two logic in sync?

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java [hunk 11]
```diff
@@ -79,13 +79,9 @@ public void testCanonicalFunctionName() {
...
-        "SELECT OrderID, Quantity,\n"
-            + "CASE\n"
-            + "    WHEN Quantity > 30 THEN 'The quantity is greater than 30'\n"
-            + "    WHEN Quantity = 30 THEN 'The quantity is 30'\n"
-            + "    ELSE 'The quantity is under 30'\n"
-            + "END AS QuantityText\n"
-            + "FROM OrderDetails");
+        "SELECT OrderID, Quantity,\n" + "CASE\n" + "    WHEN Quantity > 30 THEN 'The quantity is greater than 30'\n"
+            + "    WHEN Quantity = 30 THEN 'The quantity is 30'\n" + "    ELSE 'The quantity is under 30'\n"
+            + "END AS QuantityText\n" + "FROM OrderDetails");
```
Comment (amrishlal) [inline pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:84]: Can we avoid unrelated formatting changes from this PR, so that it is easier to see what has changed or not changed?

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java [hunk 12]
```diff
@@ -191,45 +178,148 @@ public void testQuotedStrings() {
...
-    Assert.assertEquals(func.getOperands().get(1).getLiteral().getStringValue(), "^U.*");
-    pinotQuery = CalciteSqlParser.compileToPinotQuery("select * from vegetables where g IN (12, 13, 15.2, 17)");
-    func = pinotQuery.getFilterExpression().getFunctionCall();
-    Assert.assertEquals(func.getOperator(), FilterKind.IN.name());
-    Assert.assertEquals(func.getOperands().get(0).getIdentifier().getName(), "g");
-    Assert.assertEquals(func.getOperands().get(1).getLiteral().getLongValue(), 12L);
-    Assert.assertEquals(func.getOperands().get(2).getLiteral().getLongValue(), 13L);
-    Assert.assertEquals(func.getOperands().get(3).getLiteral().getDoubleValue(), 15.2);
-    Assert.assertEquals(func.getOperands().get(4).getLiteral().getLongValue(), 17L);
+    {
```
Comment (amrishlal) [inline pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:181]: Same here. Can we avoid this formatting changes?

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java (hunk 8 above)
Comment (richardstartin) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:44]: I would prefer this were solved differently as requested previously:

```java
public enum FilterKind {
  AND(false, false),
  OR(false, false),
  NOT(false, false),
  EQUALS(true,  false),
  NOT_EQUALS(true,  false),
  GREATER_THAN(true,  true),
  GREATER_THAN_OR_EQUAL(true,  true),
  LESS_THAN(true,  true),
  LESS_THAN_OR_EQUAL(true,  true),
  LIKE(true,  false),
  BETWEEN(true,  true),
  RANGE(true,  true),
  IN(true,  false),
  NOT_IN(true,  false),
  REGEXP_LIKE(true,  false),
  IS_NULL(true,  false),
  IS_NOT_NULL(true,  false),
  TEXT_MATCH(true,  false),
  JSON_MATCH(true,  false);
  
  private final boolean _isPredicate;
  private final boolean _isRange;

  FilterKind(boolean isPredicate, boolean isRange) {
    _isPredicate = isPredicate;
    _isRange = isRange;
  }

  /**
   * Helper method that returns true if the enum maps to a Range.
   *
   * @return True if the enum is of Range type, false otherwise.
   */
  public boolean isRange() {
    return _isRange;
  }
}
```

this forces the modifier to define whether the new value is a predicate or a range predicate.

Comment (amrishlal) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:44]: Yes, this approach looks good. Sorry I missed your earlier comment.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java [hunk 13]
```diff
@@ -18,21 +18,27 @@
...
 import org.apache.pinot.pql.parsers.pql2.ast.FilterKind;
 import org.apache.pinot.sql.parsers.SqlCompilationException;
 
-
 public class PredicateComparisonRewriter implements QueryRewriter {
   @Override
   public PinotQuery rewrite(PinotQuery pinotQuery) {
     Expression filterExpression = pinotQuery.getFilterExpression();
     if (filterExpression != null) {
+      filterExpression = updateBooleanPredicates(filterExpression, null);
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java:38]: I'd do it within the `updateComparisonPredicate` (you may rename it to `updatePredicate`) so that it can be applied to both filter and having, and we just need to traverse the tree once

**Review by Jackie-Jiang — COMMENTED** [review]
You might also want to modify `RequestContextUtils.getFilter()` (note there are 2 of them with different parameters) because some predicate might be compiled at run time (e.g. aggregation filter), which is not handled in the rewriter

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java (hunk 8 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java]: Thanks for pointing this out. Done.

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java (hunk 10 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java]: Done.

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java (hunk 11 above)
Comment (vvivekiyer) [reply pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java]: Done.

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java (hunk 12 above)
Comment (vvivekiyer) [reply pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:194]: I'd prefer to keep this change. The test is cleaner this way.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java (hunk 13 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java]: The expression within the WHERE and HAVING clause should be of boolean type, and `sum(field) > 10` will return a boolean result. Some functions can also return boolean result, and is not captured as `FilterKind`, e.g. `isWholeNumber(sum(field))`. For the HAVING clause, the aggregation result is treated as a regular column, so we should handle it the same as WHERE clause.

**Review by richardstartin — APPROVED** [review]

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java [hunk 14]
```diff
@@ -33,75 +36,138 @@ public class PredicateComparisonRewriter implements QueryRewriter {
...
+   This method converts a predicate expression to the what Pinot could evaluate.
+   1. For comparison expression, left operand could be any expression, but right operand only
+   supports literal. E.g. 'WHERE a > b' will be converted to 'WHERE a - b > 0'
+   2. Updates boolean predicates (literals and scalar functions) that are missing an EQUALS filter.
+   *
+   *
+   * @param expression current expression in the expression tree
+   * @param parentFunction parent expression
+   * @return re-written expression.
+   */
```
[the remark this answers is not in the document]
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java]: Done.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java (hunk 13 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java]: Thanks for the pointer. Handled this case.

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java [hunk 15]
```diff
@@ -166,91 +168,119 @@ public static FunctionContext getFunction(FunctionCallAstNode astNode) {
...
-            new IsNullPredicate(getExpression(operands.get(0))));
-      case IS_NOT_NULL:
-        return new FilterContext(FilterContext.Type.PREDICATE, null,
-            new IsNotNullPredicate(getExpression(operands.get(0))));
-      default:
-        throw new IllegalStateException();
+    return getFilterHelper(thriftExpression, null);
+  }
+
+  private static FilterContext getFilterHelper(Expression thriftExpression, Function parentFunction) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java:174]: I don't follow the logic here. Why do we need to pass in the `parentFunction` here? We should never call this method when the parent is a predicate.

Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java:174]: I added it as a safeguard for future-proofing. It is true that we never call when parentFunction is a predicate. Hence I throw an exception when parentFunction is a predicate.
If it's not needed, I can keep the existing function definition as it is currently and make the changes. Thoughts?

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java:174]: I'd suggest not checking the parent function for the boolean rewrite. Logically, the parsing should not rely on the parent function

Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java]: Makes sense. Made the changes. Please review.

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java [hunk 16]
```diff
@@ -163,94 +165,113 @@ public static FunctionContext getFunction(FunctionCallAstNode astNode) {
...
-        return new FilterContext(FilterContext.Type.PREDICATE, null,
-            new TextMatchPredicate(getExpression(operands.get(0)), getStringValue(operands.get(1))));
-      case JSON_MATCH:
-        return new FilterContext(FilterContext.Type.PREDICATE, null,
-            new JsonMatchPredicate(getExpression(operands.get(0)), getStringValue(operands.get(1))));
-      case IS_NULL:
-        return new FilterContext(FilterContext.Type.PREDICATE, null,
-            new IsNullPredicate(getExpression(operands.get(0))));
-      case IS_NOT_NULL:
+    ExpressionType type = thriftExpression.getType();
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java:172]: Suggest refactoring the code a little bit so that it does not show as refactoring the whole thing:
- Add `private static FilterContext getFilter(Function thriftFunction)`
- Do a `switch(type)` and call it if it is `FUNCTION` type

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java [hunk 17]
```diff
@@ -163,94 +165,113 @@ public static FunctionContext getFunction(FunctionCallAstNode astNode) {
...
+        default:
+          throw new IllegalStateException();
+      }
+    } else if (type == ExpressionType.IDENTIFIER) {
+      // Convert "WHERE a" to "WHERE a = true"
+      return new FilterContext(FilterContext.Type.PREDICATE, null,
+          new EqPredicate(getExpression(thriftExpression), getStringValue(RequestUtils.getLiteralExpression(true))));
     }
+
+    throw new IllegalArgumentException();
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java:274]: We should also handle `literal`, where we handle `false`, `0`, `0.0` as false, others as true. Currently literal is always represented as string, so we have to take this work-around. Once we preserve the literal type, this can be simplified

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java [hunk 18]
```diff
@@ -19,33 +19,41 @@
...
-  BETWEEN,
-  RANGE,
-  IN,
-  NOT_IN,
-  REGEXP_LIKE,
-  IS_NULL,
-  IS_NOT_NULL,
-  TEXT_MATCH,
-  JSON_MATCH;
+  AND(false, false),
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:22]: Is this change related? We should not rely on type `FilterKind` for the purpose of this PR

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java [hunk 19]
```diff
@@ -19,33 +19,41 @@
...
+
+  /** @return true if the enum is of Range type, false otherwise. */
   public boolean isRange() {
-    return this == GREATER_THAN || this == GREATER_THAN_OR_EQUAL || this == LESS_THAN || this == LESS_THAN_OR_EQUAL
-        || this == BETWEEN || this == RANGE;
+    return _isRange;
+  }
+
+  /** @return true if the filter is a predicate. This logic should mimic FilterContext.Type. */
+  public boolean isPredicate() {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:56]: Don't add this API for now if it is not used

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java [hunk 20]
```diff
@@ -33,75 +36,126 @@ public class PredicateComparisonRewriter implements QueryRewriter {
...
             }
-          }
-          break;
+            break;
+        }
       }
+    } else if (type == ExpressionType.IDENTIFIER) {
+      expression = convertPredicateToBooleanExpression(expression, parentFunction);
     }
+
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java:128]: We should also update literal to match the SQL semantic

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java (hunk 16 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java:172]: Done. Thanks!

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java (hunk 18 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:22]: Yes, we use FilterKind in PredicateComparisonRewriter. It's used to determine we need to convert to EQUALS expression. For example:
`WHERE col1 and col2` should be converted
 `WHERE col1 BETWEEN 10 AND 20` should not be converted.

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java (hunk 19 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:56]: Please see comment above.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java (hunk 20 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java:128]: Can we add support for literals as a separate PR - asking because adding support and test for literals will make this PR larger? I'll send out a PR just for literals soon.

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java (hunk 17 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java]: Can we add support for literals as a separate PR? (asking because adding support and test for literals will make this PR larger). I'll send out a PR just for literals soon.

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java (hunk 18 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/FilterKind.java:22]: I checked the usage of `isPredicate()` and it is actually not needed. See the comments in `PredicateComparisonRewriter`.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java [hunk 21]
```diff
@@ -33,75 +36,126 @@ public class PredicateComparisonRewriter implements QueryRewriter {
...
+   *    supports literal. E.g. 'WHERE a > b' will be converted to 'WHERE a - b > 0'
+   * 2. Updates boolean predicates (literals and scalar functions) that are missing an EQUALS filter.
+   *    E.g. 1:  'WHERE a' will be updated to 'WHERE a = true'
+   *    E.g. 2: "WHERE startsWith(col, 'str')" will be updated to "WHERE startsWith(col, 'str') = true"
+   *
+   * @param expression current expression in the expression tree
+   * @param parentFunction parentFunction parent expression
+   * @return re-written expression.
+   */
+  private static Expression updatePredicate(Expression expression, Function parentFunction) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java:60]: We don't need to pass in the `parentFunction` here. The rewrite should not rely on the parent function

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java [hunk 22]
```diff
@@ -33,75 +36,126 @@ public class PredicateComparisonRewriter implements QueryRewriter {
...
-            if (!secondOperand.isSetLiteral()) {
-              function.setOperator(getOppositeOperator(filterKind).name());
-              operands.set(0, secondOperand);
-              operands.set(1, firstOperand);
+      if (!EnumUtils.isValidEnum(FilterKind.class, functionOperator)) {
+        // If the function is not of FilterKind, we might have to rewrite the function if parentFunction is not a
+        // predicate.
+        // Example: A query like "select col1 from table where startsWith(col1, 'myStr') AND col2 > 10;" should be
+        //          rewritten to "select col1 from table where startsWith(col1, 'myStr') = true AND col2 > 10;".
+        expression = convertPredicateToBooleanExpression(expression, parentFunction);
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java:71]: The example here does not apply because `AND` is a valid `FilterKind`, and won't get into this branch.
This can be simplified to just rewriting the expression to `expression = true` without any additional check

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java (hunk 20 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java:128]: Sounds good, we may add a TODO here and handle it separately

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java [hunk 23]
```diff
@@ -33,75 +36,126 @@ public class PredicateComparisonRewriter implements QueryRewriter {
...
+   *               "select * from table where startsWith(col1, 'str') = true"
+   *  2. Separator is not a predicate (non predicates are AND, OR, NOT).
+   *     Example1: "select * from table where col1 AND startsWith(col2, 'str')" converts to
+   *               "select * from table where col1 = true AND startsWith(col2, 'str') = true"
+   * @param expression Expression
+   * @param parentFunction Parent expression
+   * @return Rewritten expression
+   */
+  private static Expression convertPredicateToBooleanExpression(Expression expression, Function parentFunction) {
+    if (parentFunction == null || (EnumUtils.isValidEnum(FilterKind.class, parentFunction.getOperator()) && !FilterKind
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java:147]: This condition should always be met, and we should be able to rewrite the expression without any check. `parentFunction` is not needed here

Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java]: I see your point now. The code is much cleaner following this suggestion. Thanks!

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java (hunk 21 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java]: Done.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java (hunk 22 above)
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/PredicateComparisonRewriter.java]: Done.

**Review by Jackie-Jiang — APPROVED** [review]
LGTM
